llama-index-core/llama_index/core/memory/vector_memory.py

```diff
@@ -0,0 +1,234 @@
+        return cls(vector_index=index, retriever_kwargs=retriever_kwargs)
+
+    def get(
+        self, input: Optional[str] = None, initial_token_count: int = 0, **kwargs: Any
+    ) -> List[ChatMessage]:
+        """Get chat history."""
+        if input is None:
+            raise ValueError("Input must be provided to get chat history.")
+
+        # retrieve from index
+        retriever = self.vector_index.as_retriever(**self.retriever_kwargs)
+        nodes = retriever.retrieve(input or "")
+
+        # retrieve underlying messages
+        messages = [
+            ChatMessage.parse_obj(sub_dict)
+            for node in nodes
+            for sub_dict in node.metadata["sub_dicts"]
+        ]
+
+        # add system message
+        if self.system_message:
+            messages = [
+                ChatMessage.from_str(self.system_message, role=MessageRole.SYSTEM)
+            ].extend(messages)
+
+        if self.return_single_message:
+            # condense all messages into a single message
+            messages = [
+                ChatMessage.from_str(
+                    " ".join([m.content for m in messages]), role=MessageRole.USER
+                )
+            ]
+
+        return messages
+
+    def get_all(self) -> List[ChatMessage]:
+        """Get all chat history."""
+        # TODO: while we could implement get_all, would be hacky through metadata filtering
+        # since vector stores don't easily support get()
+        raise ValueError(
+            "Vector memory does not support get_all method, can only retrieve based on input."
+        )
+
+    def _commit_node(self, override_last: bool = False) -> None:
+        """Commit new node to vector store."""
+        # commit to vector store
+        node_id = str(uuid.uuid4())
+        # create subnodes for each message
+        sub_dicts = []
+        for msg in self.chat_store.get_messages(self.cur_user_msg_key):
+            sub_dicts.append(_stringify_chat_message(msg))
+
+        if not sub_dicts:
+            return
+
+        # now create a "super" node that contains all subnodes as metadata
+        # this metadata is excluded from embedding and LLM synthesis
+        # the concatenated text is put into the super node text field
+        super_node = TextNode(
+            text=" ".join(
+                [str(sub_dicts[i]["content"]) for i in range(len(sub_dicts))]
+            ),
+            id_=node_id,
+            metadata={"sub_dicts": sub_dicts},
+            excluded_embed_metadata_keys=["sub_dicts"],
+            excluded_llm_metadata_keys=["sub_dicts"],
+        )
+
+        if override_last:
+            # delete the last node
+            last_node_id = self.chat_store.delete_last_message(
+                self.chat_store_key
+            ).content
+            self.vector_index.delete_nodes([last_node_id])
+
+        self.vector_index.insert_nodes([super_node])
+        self.chat_store.add_message(self.chat_store_key, ChatMessage(content=node_id))
```

Comment:
A little confused on the flow here

The chat store is keeping track of the inserted node ids (effectively maintaining order)? Not 100% sure what override last is for

Reply:
so I think `override_last` is needed since we're updating the "batch" as we go. So if its not the end of the batch, we gotta delete the last entered node, and replace it in favor with the newly updated one... i.e. one that has the latest message.

Reply:
The `chat_store` is acting as a temp buffer (I think you had mentioned before, and I'm just getting caught up with now) to track the current user batch i.e., group.

Reply:
see working PR here for removing chat_store as buffer: https://github.com/run-llama/llama_index/pull/13729